### PR TITLE
Platform agnostic file functions

### PR DESCRIPTION
**Pull request by mcostalba**
Reduce the platform specific code to bare minimum
and localize all in a single place.

### CODE REVIEW

**Comment by lucasart** [comment]
Hum... I don't understand what the compiler is complaining about:
```
g++ -Wall -Wcast-qual -fno-exceptions -fno-rtti -std=c++11  -pedantic -Wextra -Wshadow -m64 -DNDEBUG -O3 -DIS_64BIT -msse -msse3 -mpopcnt -DUSE_POPCNT -DUSE_PEXT -mbmi -mbmi2 -flto   -c -o syzygy/tbprobe.o syzygy/tbprobe.cpp
In file included from syzygy/tbprobe.cpp:20:0:
syzygy/tbcore.cpp: In member function ‘char* TBFile::map(uint64_t*)’:
syzygy/tbcore.cpp:77:42: error: invalid conversion from ‘int’ to ‘std::ios_base::openmode {aka std::_Ios_Openmode}’ [-fpermissive]
     int fd = open(fname.c_str(), O_RDONLY);
                                          ^
```

**Comment by mcostalba** [comment]
It seems a name resolution issue, fstream has a member called 'open', try
this:

int fd = ::open(fname.c_str(), O_RDONLY);


Oddly enough I don't get the compile error on my platform.

**Comment by stuwph** [comment]
syzygy access seems to be broken, even after 'additional formatting' 
(windows x64 mingw 5.30 compile)

**Comment by stuwph** [comment]
still no syzygy after 'assorted code reshuffling' under windows fritz15 and arena351 gui
initializing the tables, but while accessing in play throwing 'corrupted table'  (arena)

**Comment by stuwph** [comment]
worked until  'Platform agnostic file functions'

File: src/syzygy/tbcore.cpp [hunk 1]
```diff
@@ -466,21 +226,125 @@ static const short KK_idx[10][64] = {
...
+public:
+    // Open the file with the given name found among the TBPaths
+    TBFile(const std::string& f) : fname(f) {
+
+        std::stringstream ss(TBPaths);
+        std::string path;
+
+        while (std::getline(ss, path, SEP_CHAR)) {
+            std::ifstream::open(path + "/" + fname);
+            if (is_open())
```
Comment (lucasart) [inline src/syzygy/tbcore.cpp]: before `return`, you need to set `fname = path + "/" + fname`

File: src/syzygy/tbcore.cpp [hunk 2]
```diff
@@ -466,21 +226,125 @@ static const short KK_idx[10][64] = {
...
+
+        close();
+
+#ifndef _WIN32
+        struct stat statbuf;
+        int fd = ::open(fname.c_str(), O_RDONLY);
+        fstat(fd, &statbuf);
+        *mapping = statbuf.st_size;
+        char* data = (char*)mmap(NULL, statbuf.st_size, PROT_READ, MAP_SHARED, fd, 0);
+        close(fd);
```
Comment (lucasart) [inline src/syzygy/tbcore.cpp]: Same as `open()`, you need `::close(fd)`

**Comment by lucasart** [comment]
@mcostalba: i managed to compile it, and fix the obvious. see comments in line. but it is still broken in a more subtle way now... wrong syzygy signature.

**Comment by mcostalba** [comment]
@lucasart I have fixed the issues and updated the pull request. Can you please confirm that is broken on [this](https://github.com/official-stockfish/Stockfish/pull/643/commits/0090c72e356314da) commit?

**Comment by mcostalba** [comment]
I have finally found why I crash on sygyzy.epd, here there are some illegal mate positions:

K7/8/1q6/8/6k1/8/8/8 w - - 71
6b1/8/8/2b5/8/7K/6Q1/6k1 b - - 46
k2q1Kr1/4n3/8/3b4/8/8/8/8 w - - 83
8/7k/8/8/R7/8/2K4Q/1R4Q1 b - - 5
3K4/1b3q2/3b4/8/2k5/2P5/8/8 w - - 43

And there are many more, without these I am able to run the bench with TB.

@lucasart how you are able to run the bench on those? each of these positions fail on SF even without TB !

**Comment by mcostalba** [comment]
OK, now that I am able to run 'bench' I have finally found the bug: PR fixed and ready to be merged.

**Comment by lucasart** [comment]
@mcostalba: Ah, I fixed my random position generator to filter mates and stalemates, but did not update `syzygy.epd`... I will open a pull request to update `syzygy.epd`. SF can run on these positions, but not in debug mode (assert crash).

**Comment by lucasart** [comment]
@mcostalba: patch is ok to merge.

**Comment by stuwph** [comment]
getting a mingw 5.3 linker error since 'Improve TBFile API'
 
syzygy/tbprobe.o:tbprobe.cpp:(.text.unlikely+0x24e7): undefined reference to `(anonymous namespace)::WDLEntry::TB_MAGIC'
syzygy/tbprobe.o:tbprobe.cpp:(.text.unlikely+0x3472): undefined reference to `(anonymous namespace)::DTZEntry::TB_MAGIC'

**Comment by mcostalba** [comment]
Which version of gcc you have?

Type g++ - v

**Comment by stuwph** [comment]
thread model: posix
gcc version 5.3.0 (Rev5, Built by MSYS2 project)

**Comment by mcostalba** [comment]
Ok, I have pushed a possible fix, could you please verify now it works?

**Comment by stuwph** [comment]
works again :)

**Pull request merged**